rfcs/2021-11-22-design-tokens.md

```diff
@@ -0,0 +1,106 @@
+
+## Scope/Outline
+
+**An outline for how this process could proceed:**
+
+1. The SSOT should be decided on, at the moment I see two options:
+
+   1. **The DSM becomes the source of truth**. Designers and developers work together to consolidate the current set of design rules residing in the codebase/Sketch and place them into the DSM. In this case, with that work done, an application can be written in the design system to utilize the [DSM Token API](https://support.invisionapp.com/hc/en-us/articles/360049950931) to periodically update it's set of Sass variables with data from these design tokens and store them in a new NPM package.
+   2. **Tokens are modified directly as sets of JSON files**. Designers and developers work together to consolidate tokens in a hierarchy which makes the most sense for all platforms. These tokens could be stored in a new NPM package. Changes to tokens would require a pull request and have a release cadence.
+
+2. Once the SSOT is determined, tokens should be created in the appropriate format. At minimum, tokens should also be populated to form a set required to serve the current consuming applications. Ongoing work could then be ticketed to fill out the entire token library. These first tokens would represent the most basic items such as spacing, color, type attributes, breakpoints, animation easing values, border thickness, etc.. An example of a color token setup might resemble:
+
+   ```
+   tokens.git/core/colors.json
+   {
+   	"lightblue": {
+   		"100": "rgb(207, 239, 252)",
+   		"200": "rgb(113, 202, 243)",
+   		"300": "rgb(43, 199, 252)",
+   		"transparent": {
+   			"50": "rgba(43, 199, 252, 50)",
+   			"75": "rgba(43, 199, 252, 75)"
+   		}
+   	}
+   }
+   ```
+
+   > The aliases design tokens are mapped to in the various consuming systems need not match up directly one another initially. The token is a standard atomic unit which can be referenced in many systems using many aliases but should be applied as high up in the dependency tree as possible.
+
+3. When the first dataset is completed and ready for use, a tool stored in the NPM package would need to be created to convert these tokens into new formats, initially Sass. There are several existing tools which can do this such as [Style Dictionary](https://github.com/amzn/style-dictionary) (format is [easily utilized by the DSM](https://support.invisionapp.com/hc/en-us/articles/360049483032)), [Theo](https://github.com/salesforce-ux/theo), [JSON -> CSS](https://github.com/rlapoele/json-to-scss) and others.
+
+   I believe the best option would be to write our own [Webpack Loader](https://webpack.js.org/contribute/writing-a-loader/) to convert these tokens to top level Sass files.
```

Comment:
Would we have to change our build tools in order to utilize this custom loader since I don't believe our current build tools use Webpack (aside from Storybook)

Reply:
I'm envisioning the tokens being in their own npm package with custom tooling. I have an interest in moving our system [off gulp and onto webpack](https://jira.cms.gov/browse/WNMGDS-1239) so it should slot in nicely over the next.. 6 months?

Reply:
Could you explain how this would work a little more? Would this loader only be used in an intermediary stage where the token files exist in our core package and are not compiled before the design system's own build step? Would there be an import in our Sass files that looks something like
```scss
@import "../tokens/colors.json";
```
and the loader would basically convert `colors.json` into a Sass file that has Sass variables?

Reply:
What benefits does writing our own converter have over the previously mentioned options? Is there a maintenance cost on our end for this custom tool? 

My thinking here is that we encountered problems using our custom built docs site to build our components (custom solution) and now we're migrating to Storybook because it's an industry standard and we don't have to spend time maintaining it (third party solution). Is there a risk we end up doing the same thing with tokens?

Reply:
@pwolfert yes essentially the loader just converts the json to sass variables we can reference in our DS variable files. This gives us the option to gradually replace the currently used vars with token values. Those variable names would match up to the naming convention designers use in sketch and anyone else utilizing our token library.

@zarahzachz the converter would be very basic for the DS, simply giving us more Sass vars to work with, but those vars and naming conventions would be unified with anyone else using the token library in name/value. The conversion would be a very simple set of transforms. This change shouldn't effect application downstream and just gives us a palette of pre-defined, commonly structured values to work with.

Reply:
Does that make sense? Should I expand that to make it more clear?